### ISSUE TITLE
[YSQL][Index]: Encountering ERROR:  Aborted: ERROR:  Query error: schema version mismatch for table 0000562900003000800000000000563c: expected 14, got 13 while trying to create index for a table with locked row

### ISSUE THREAD

**Issue by Karvy-yb** [yugabyte/yugabyte-db#14222]
Jira Link: [DB-3683](https://yugabyte.atlassian.net/browse/DB-3683)
### Description

Steps to reproduce:

1. Open shell session (s1)
2. s1: `create table t1(id int primary key, age int);`
3. s1: `insert into t1 values(1, 11);`
4. s1: `insert into t1 values(2, 12);`
5. Open another shell session (s2) and run `begin;`
6. s2: `select * from t1 where id=1 for update;`
7. s1: `create index t1_ind1 on t1(age);` 

Step 7 will give the `ERROR:  Aborted: ERROR:  Query error: schema version mismatch for table 00005629000030008000000000005657: expected 1, got 0` message. 
If I try to recreate the index then I get `ERROR:  relation "t1_ind1" already exists` message as the index is created, shown in the UI of platform but is in invalid state.

Note:
Same error is observed when the row level lock is acquired non transactionally.

Edit: Observed this on universe with pessimistic locking disabled as well. It is reproducible.

**Comment by jasonyb** [comment]
Next time, please try to find a minimal repro when you have the time.  For example, you don't need row locking and you don't need two inserts.

Here is a smaller repro:

A: `CREATE TABLE t (i int);`
B: `SELECT * FROM t;`
A: `CREATE INDEX ON t (i);`

A must be on different node from B.  Table t must have a tablet on B's node (should happen by default on rf-3 3-master 3-tserver cluster).

This is solved by commit 7583bcc0ea272ef8742adff796b36e354f9b6767.  On recent master commit 09b7303ad4b45c68c65090cbbb09897d66ee4add, applying the key revert

```diff
diff --git i/src/yb/tserver/tablet_server.cc w/src/yb/tserver/tablet_server.cc
index 75cacbdec0..84fc9e8fcc 100644
--- i/src/yb/tserver/tablet_server.cc
+++ w/src/yb/tserver/tablet_server.cc
@@ -638,7 +638,7 @@ void TabletServer::SetYsqlCatalogVersion(uint64_t new_version, uint64_t new_brea
   auto pg_client_service = pg_client_service_.lock();
   if (pg_client_service) {
     LOG(INFO) << "Invalidating PgTableCache cache since catalog version incremented";
-    pg_client_service->InvalidateTableCache();
+//    pg_client_service->InvalidateTableCache();
   }
 }

```

I can reproduce the issue with my small repro and your repro (provided s1 and s2 are different nodes).  Without this revert, there is no issue.

The reason is that a new pg session would not invalidate tserver's pg table cache.  In my repro, B caches t's v0 in tserver pg table cache.  CREATE INDEX increments it to v1 then launches backfill, which involves spawning pg sessions to read from the table (to do backfill).  These sessions won't invalidate tserver cached v0, so they read the old v0 value and use that in read request, which mismatches with tserver's v1.

A similar but different repro:

A: `CREATE TABLE t (i int);`
B: `SELECT * FROM t; -- cache v0`
B: `exit;`
A: `ALTER TABLE t ADD CHECK (i > 0); -- bump to v1`
B: `SET yb_debug_log_internal_restarts = ON;`
B: `SELECT * FROM t; -- load v0, get mismatch, then transparent retry (see logs)`

Closing since this is fixed in master, but it may not be fixed in all stable branches: @sanketkedia, please check.

cc: @d-uspenskiy

**Comment by jasonyb** [comment]
The commit 7583bcc0ea272ef8742adff796b36e354f9b6767 fix is sufficient for CREATE INDEX:

1. table schema version is incremented
1. indislive=t is committed, leading to catalog version bump
1. catalog version bump causes tservers to invalidate pg table cache (because of above mentioned commit)
1. indisready=t is committed
1. backfill happens, and pg processes spawned by that will no longer read old table schema cached in tserver since that was invalidated

There is no room for issue in case of different timing besides the overall issue of catalog version not propagating promptly as we would expect, but this is a general issue for all DDLs (part 1 of issue #13369).

The issue described here is still applicable to other operations like a COPY on a new session since COPY has no internal retry.

A: `CREATE TABLE t (i int);`
B: `SELECT * FROM t; -- cache v0`
B: `exit;`
A: `ALTER TABLE t ADD CHECK (i > 0); -- bump to v1`
B: `COPY t TO '/dev/null';`

There is room for timing to cause the COPY to still hit the issue even with the commit 7583bcc0ea272ef8742adff796b36e354f9b6767 fix if COPY happens between the ALTER's docdb table schema change and the postgres sys catalog change (docdb table schema change is used for the schema version check, and postgres sys catalog change increments catalog version that triggers tserver pg table cache invalidation).  The order of operations of ALTER may change by the DDL atomicity project, cc: @deeps1991.

I do not create a dedicated issue for those cases since schema version mismatch is not uncommon and can already happen easily for other reasons.  For the above example, by moving the ALTER to the middle of a long COPY, schema version mismatch would happen anyway.

Also, in case it wasn't clear here and the previous post, `exit;` is necessary because

- without exit, before starting the next query after ALTER happened, a catalog version bump is detected, and that results in deep cache refresh (invalidates tserver pg table cache).
- with exit, it initializes with up-to-date catalog version, so no invalidation is requested.
